src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file copy_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_COPY_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "copy.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+CopyTask::CopyTask(const size_t maxLength,
+                   const size_t nRepeats,
+                   const bool addSeparator) :
+    maxLength(maxLength),
+    nRepeats(nRepeats),
+    addSeparator(addSeparator)
+{
+  if (maxLength <= 1)
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): maximum sequence length ("
+        << maxLength << ") "
+        << "should be at least 2!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  if (nRepeats <= 0)
+  {
+    std::ostringstream oss;
+    oss << "CopyTask::CopyTask(): repetition count (" << nRepeats << ") "
+        << "is not positive!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+  // Just storing task-specific parameters.
+}
+
+const void CopyTask::Generate(arma::field<arma::mat>& input,
+                              arma::field<arma::mat>& labels,
+                              const size_t batchSize,
+                              bool fixedLength)
+{
+  input = arma::field<arma::mat>(batchSize);
+  labels = arma::field<arma::mat>(batchSize);
+  size_t size = maxLength;
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    if (!fixedLength)
+    {
+      arma::vec weights(maxLength - 1);
+      // We have two binary numbers with exactly two digits (10 and 11).
+      // Increasing length by 1 double the number of valid numbers.
+      weights = arma::exp2(arma::linspace(1, maxLength - 1, maxLength - 1));
+
+      size = 2 + mlpack::math::RandInt(weights);
+    }
+    arma::colvec vecInput = arma::randi<arma::colvec>(
+      size, arma::distr_param(0, 1));
+    arma::colvec vecLabel = arma::conv_to<arma::colvec>::from(
+      arma::repmat(vecInput, nRepeats, 1));
+    size_t totSize = vecInput.n_elem + addSeparator + vecLabel.n_elem;
+    input(i) = arma::zeros(totSize, 2);
+    input(i).col(0).rows(0, vecInput.n_elem-1) =
+      vecInput;
```

Comment:
Some small style issues here---the wrapped lines should be indented 4 spaces, not 2, and `vecInput.n_elem-1` should be `vecInput.n_elem - 1`. :)